### PR TITLE
[SPARK-37267][SQL] OptimizeSkewInRebalancePartitions support optimize non-root node

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Make `OptimizeSkewInRebalancePartitions` rule support optimize non-root node.

### Why are the changes needed?

`OptimizeSkewInRebalancePartitions` now is only applied if the `RebalancePartitions` is the root node, but sometimes, we expect a local sort after do RebalancePartitions that can improve the compression ratio.

For example:
```sql
SELECT /*+ REBALANCE(c1) */ * FROM t SORT BY c1
```

After SPARK-36184, we make validate easy that whether the rule introduces extra shuffle or not and the output partitioning is ensured by `AQEShuffleReadExec.outputPartitioning`.

So it is safe to make `OptimizeSkewInRebalancePartitions` support optimize non-root node.

### Does this PR introduce _any_ user-facing change?

maybe yes, if user do some operator after `RebalancePartitions`, the shuffle reader will be changed.

### How was this patch tested?

add new test in `org.apache.spark.sql.execution.adaptive.AdaptiveQueryExecSuite`

### CODE REVIEW

**Comment by ulysses-you** [comment]
retest this please

**Comment by ulysses-you** [comment]
cc @yaooqinn @cloud-fan @HyukjinKwon @wangyum

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 1]
```diff
@@ -2192,6 +2192,37 @@ class AdaptiveQueryExecSuite
...
+    withTempView("v") {
+      withSQLConf(
+        SQLConf.ADAPTIVE_OPTIMIZE_SKEWS_IN_REBALANCE_PARTITIONS_ENABLED.key -> "true",
+        SQLConf.SHUFFLE_PARTITIONS.key -> "1",
+        SQLConf.COALESCE_PARTITIONS_MIN_PARTITION_NUM.key -> "1") {
+        spark.sparkContext.parallelize(
+          (1 to 10).map(i => TestData(if (i > 2) 2 else i, i.toString)), 2)
+          .toDF("c1", "c2").createOrReplaceTempView("v")
+
+        def checkSkew(query: String, skewedPartitionNumber: Int): Unit = {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:2206]:
```suggestion
        def checkRebalance(query: String, numShufflePartitions: Int): Unit = {
```

**Review by cloud-fan — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 2]
```diff
@@ -2192,6 +2192,37 @@ class AdaptiveQueryExecSuite
...
+        def checkRebalance(query: String, numShufflePartitions: Int): Unit = {
+          val (_, adaptive) = runAdaptiveAndVerifyResult(query)
+          assert(adaptive.collect {
+            case sort: SortExec => sort
+          }.size == 1)
+          val read = collect(adaptive) {
+            case read: AQEShuffleReadExec => read
+          }
+          assert(read.size == 1)
+          assert(read.head.partitionSpecs.count(_.isInstanceOf[PartialReducerPartitionSpec]) ==
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:2215]: In this test, I think all the partitionSpecs are `PartialReducerPartitionSpec`?

Comment (ulysses-you) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:2215]: yes, added the assert

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:2215]: then should this be `assert(read.head.partitionSpecs.forall(_.isInstanceOf[PartialReducerPartitionSpec])`?

Comment (ulysses-you) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:2215]: updated

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by ulysses-you** [comment]
thank you @cloud-fan @yaooqinn

**Pull request closed without merging**